src/components/AnchoredTextLabels.cpp

```diff
@@ -0,0 +1,61 @@
+#include "AnchoredTextLabels.h"
+#include <Components/BsCCamera.h>
+#include <GUI/BsCGUIWidget.h>
+#include <GUI/BsGUILabel.h>
+#include <GUI/BsGUIPanel.h>
+#include <Math/BsMath.h>
+#include <Scene/BsSceneObject.h>
+#include <excepction/Throw.hpp>
+
+namespace REGoth
+{
+  AnchoredTextLabels::AnchoredTextLabels(const bs::HSceneObject& parent)
+      : Component(parent)
+  {
+    // Set a name for the component, so we can find it later if needed
+    setName("AnchoredTextLabels");
+
+    // A GUI widget component is required for drawing the labels
+    mGui = parent->getComponent<bs::CGUIWidget>();
+
+    if (mGui.isDestroyed())
+    {
+      REGOTH_THROW(
+          InvalidStateException,
+          bs::StringUtil::format("Scene Object {0} does not have a CGUIWidget component!",
+                                 SO()->getName()));
+    }
+  }
+
+  void AnchoredTextLabels::addLabel(const bs::Vector3& anchorPosition, const bs::HString& text)
+  {
+    bs::GUILabel* label = bs::GUILabel::create(text);
+    mLabels.emplace_back(AnchoredLabel{anchorPosition, label});
+
+    bs::GUIPanel* mainPanel = mGui->getPanel();
+    mainPanel->addElement(label);
+  }
+
+  void AnchoredTextLabels::update()
+  {
+    auto camera                          = mGui->getCamera();
+    const bs::Transform& cameraTransform = camera->getTransform();
+    const bs::Vector3& cameraPosition    = cameraTransform.getPosition();
+    const bs::Vector3& cameraForward     = cameraTransform.getForward();
+    for (const auto& anchoredLabel : mLabels)
+    {
+      // Only show labels with an anchor point in front of the camera
+      if ((anchoredLabel.mWorldPosition - cameraPosition).dot(cameraForward) > 0.f)
```

Comment:
Could be extracted into a function `isPointInFrontOfCamera` or so.

Reply:
I would expect such a function to take a camera and a position as input. Unfortunately that wouldn't allow me to get the camera position and forward direction only once beforehand but I would have to do it for every label. I mean it's some kind of premature optimization but I thought this way would be the most future proof one.

By the way, thinking of this line I had the idea that it would probably be helpful to specify something like a maximum distance. Like "only draw these labels if their anchor points are closer than *x* to the camera". Could be easily done as the dot product gives us the distance (assuming `cameraForward` is normalized). What do you thinkg about it?

Reply:
I don't remember how Gothic did it, but right now all the text would be visible through walls, so adding a max-distance would be the easiest solution to not clutter the screen now.

As for optimization, just pass a reference to the cameras transform. The compiler should be smart enough to optimize the calls to `getForward()` etc. It would make adding the distance check cleaner as well IMHO.

Reply:
I added the distance check and separate function in 6889ece. We could potentially hide text labels if their anchor point is behind world geometry by looking at the depth buffer of the attached camera. But I don't know how and if that is possible and it would probably be an overkill ATM.